### PR TITLE
LibTTF: Finish work started by Srimanta Barua

### LINKED ISSUE #743: LibGUI: Implement TrueType and/or OpenType fonts
Currently Serenity only supports pixel based fonts, and as such are limited in size and cannot be scaled. (unless of course you did some type of pixel based scaling with anti-aliasing, but that would probably look terrible). On higher resolution screens with higher pixel densities the fonts also become unreadable, so implementing this would also allow for possible display scaling in the future.

The browser would also benefit greatly from this because it's currently limited to the sizes provided by the existing fonts.

### PR DESCRIPTION
**Pull request by sunverwerth**
Screenshot von 2020-12-29 16 16 47

Closes #2512
Fixes #743

I rebased this on the most recent master and got it back to working.

ToDos:

- [x] ~~Abstract font interface for TTF and pixel fonts~~
- [x] Use ReadonlyBytes to avoid unnecessary buffer copies in the TTF::Font class
- [x] Cache rendered glyphs in the TTF::ScaledFont class

I'll leave it at that for now. TTF system fonts and font viewer apps are probably out of scope for this PR.

### CODE REVIEW

File: Libraries/LibTTF/Cmap.cpp [hunk 1]
```diff
@@ -90,8 +90,7 @@ Optional<Cmap::Subtable> Cmap::subtable(u32 index) const
     u16 encoding_id = be_u16(m_slice.offset_pointer(record_offset + (u32)Offsets::EncodingRecord_EncodingID));
     u32 subtable_offset = be_u32(m_slice.offset_pointer(record_offset + (u32)Offsets::EncodingRecord_Offset));
     ASSERT(subtable_offset < m_slice.size());
-    // HACK: added const_cast because of new ByteBuffer::wrap behavior, should probably find another workaround
-    auto subtable_slice = ByteBuffer::wrap(const_cast<u8*>(m_slice.offset_pointer(subtable_offset)), m_slice.size() - subtable_offset);
+    auto subtable_slice = ByteBuffer::copy(m_slice.offset_pointer(subtable_offset), m_slice.size() - subtable_offset);
```
Comment (alimpfard) [inline Libraries/LibTTF/Cmap.cpp:93]: (if wrap was fine, you can definitely take `ReadonlyBytes` instead of copying)

Comment (alimpfard) [reply Libraries/LibTTF/Cmap.cpp:93]: I noticed that you had it as one of the TODOs, sorry for not looking harder lol

Comment (sunverwerth) [reply Libraries/LibTTF/Cmap.cpp:93]: Ah cool, that's an alias for Span<const u8>?

Comment (alimpfard) [reply Libraries/LibTTF/Cmap.cpp:93]: yep.

Comment (sunverwerth) [reply Libraries/LibTTF/Cmap.cpp]: Well, that was easy

**Comment by tomuta** [comment]
Without having looked at the actual changes, I think we should get this merged sooner than later, even if it has imperfections. We can always improve upon it iteratively. Don't want it to die like the last PR ;)

**Comment by sunverwerth** [comment]
@tomuta I'm determined to finish this now :smile: If not today, then tomorrow.

**Comment by sunverwerth** [comment]
I couldn't help myself and test ttf as a system font. The results are... interesting :laughing: 

image

Clearly, there is a lot more work to do to transition from fixed dimension bitmaps to per character sizing.
I'm on my way there but there's still a lot to do.
I would like to split up the rest of the work into a seperate PR and merge what I have so far.

So, only the LibTTF and the last commit (move default_XXX_font() into FontDatabase)

Any objections?

**Comment by sunverwerth** [comment]
@awesomekling I think you clicked edit instead of reply

Ok, I'll fix the linter issues.

Are there any change requests for the code in LibTTF itself?
I haven't touched most of it since the original PR.
Maybe someone can have a quick look. @alimpfard ?

**Comment by alimpfard** [comment]
I shall review a bit then :)

**Comment by awesomekling** [comment]
> @awesomekling I think you clicked edit instead of reply

Whoa oops, sorry about that!

**Comment by alimpfard** [comment]
> I couldn't help myself and test ttf as a system font. The results are... interesting laughing
> 
> Clearly, there is a lot more work to do to transition from fixed dimension bitmaps to per character sizing.


I think those artifacts are mainly because it's not respecting kerning (or maybe GPOS, if that's in the font).
I had kerning in my old ttf branch, but I don't think it looked anywhere near good enough with just kerning either...

**Comment by sunverwerth** [comment]
> > I couldn't help myself and test ttf as a system font. The results are... interesting laughing
> > Clearly, there is a lot more work to do to transition from fixed dimension bitmaps to per character sizing.
> 
> I think those artifacts are mainly because it's not respecting kerning (or maybe GPOS, if that's in the font).
> I had kerning in my old ttf branch, but I don't think it looked anywhere near good enough with just kerning either...

The main issue is that all text rendering functions until now expect all character bitmaps to be the same height.

**Comment by awesomekling** [comment]
Would be good to check that each commit still makes sense. This one looks kinda unnecessary for example: `LibGfx: Clean up Painter and remove duplicated methods for drawing TTF text.`

**Comment by alimpfard** [comment]
So far as the contents go, I don't think I have any nits to pick :)

**Comment by sunverwerth** [comment]
I squashed the one commit and edited the commit messages that were a bit too long. Waiting for the CI to complete.

**Pull request merged**